demos/text_detection_demo/src/text_detection.cpp

```diff
@@ -198,42 +227,64 @@ std::vector<cv::RotatedRect> postProcess(const InferenceEngine::BlobMap &blobs,
-
-    cv::Mat mask = decodeImageByJoin(cls_data, new_cls_data_shape, link_data, new_link_data_shape, cls_conf_threshold, link_conf_threshold);
-    std::vector<cv::RotatedRect> rects = maskToBoxes(mask, static_cast<float>(kMinArea),
-                                                     static_cast<float>(kMinHeight), image_size);
-
-    return rects;
+    if (!kLocOutputName.empty() && !kClsOutputName.empty()) {
+        // PostProcessing for PixelLink Text Detection model
+    	auto link_shape = blobs.at(kLocOutputName)->getTensorDesc().getDims();
+    	size_t link_data_size = link_shape[0] * link_shape[1] * link_shape[2] * link_shape[3];
+    	InferenceEngine::LockedMemory<const void> locOutputMapped = InferenceEngine::as<InferenceEngine::MemoryBlob>(
+    	    blobs.at(kLocOutputName))->rmap();
+    	float *link_data_pointer = locOutputMapped.as<float *>();
+    	std::vector<float> link_data(link_data_pointer, link_data_pointer + link_data_size);
+    	link_data = transpose4d(link_data, ieSizeToVector(link_shape), {0, 2, 3, 1});
+    	softmax(&link_data);
+    	link_data = sliceAndGetSecondChannel(link_data);
+    	std::vector<int> new_link_data_shape(4);
+    	new_link_data_shape[0] = static_cast<int>(link_shape[0]);
+    	new_link_data_shape[1] = static_cast<int>(link_shape[2]);
+    	new_link_data_shape[2] = static_cast<int>(link_shape[3]);
+    	new_link_data_shape[3] = static_cast<int>(link_shape[1]) / 2;
+
+    	auto cls_shape = blobs.at(kClsOutputName)->getTensorDesc().getDims();
+    	size_t cls_data_size = cls_shape[0] * cls_shape[1] * cls_shape[2] * cls_shape[3];
+    	InferenceEngine::LockedMemory<const void> clsOutputMapped = InferenceEngine::as<InferenceEngine::MemoryBlob>(
+    	    blobs.at(kClsOutputName))->rmap();
+    	float *cls_data_pointer = clsOutputMapped.as<float *>();
+    	std::vector<float> cls_data(cls_data_pointer, cls_data_pointer + cls_data_size);
+    	cls_data = transpose4d(cls_data, ieSizeToVector(cls_shape), {0, 2, 3, 1});
+    	softmax(&cls_data);
+    	cls_data = sliceAndGetSecondChannel(cls_data);
+    	std::vector<int> new_cls_data_shape(4);
+    	new_cls_data_shape[0] = static_cast<int>(cls_shape[0]);
+    	new_cls_data_shape[1] = static_cast<int>(cls_shape[2]);
+    	new_cls_data_shape[2] = static_cast<int>(cls_shape[3]);
+    	new_cls_data_shape[3] = static_cast<int>(cls_shape[1]) / 2;
+
+    	cv::Mat mask = decodeImageByJoin(cls_data, new_cls_data_shape, link_data, new_link_data_shape, cls_conf_threshold, link_conf_threshold);
+    	std::vector<cv::RotatedRect> rects = maskToBoxes(mask, static_cast<float>(kMinArea),
+    	                                                 static_cast<float>(kMinHeight), image_size);
+    	return rects;
+
+    } else {
+        // PostProcessing for Horizontal Text Detection model
+    	for (const auto &blob : blobs) {
+    	    if (blob.second->getTensorDesc().getDims()[1] == 5) {
+    	        kLocOutputName = blob.first;
+    	        kClsOutputName = "labels";
```

Comment:
I just noticed that `kClsOutputName` isn't used at all in this branch. Is there even a need to set it?

Reply:
This is a hack to make running the demo with other text detection models safer. In case of running unsupportable model, the script will go in the `if (kLocOutputName.empty() || kClsOutputName.empty())` condition and throw the exception.
Indeed, I don't use `kClsOutputName` in the script, but removing it will lead to simplifying the condition above to 
`if (kLocOutputName.empty())`. 
And I'm convinced that some unsupportable model may satisfy the condition, go further and crash in an unpredictable place.
I can remove it if my fears are unfounded.

Reply:
The only requirement the else branch has is that blobs' second dimensions size is 5. So once it is satisfied you can process it and be sure you won't try reading memory that doesn't belong to the blob.

But if your intention was to allow only processing of models that has exactly the same output interface as `handwritten-score-recognition-0001` than you should have checked that the blob that was going to correspond to `kClsOutputName` has only one dimension and and the size of that dimension matches with `kLocOutputName`.

While it would be nice to have such accurate and precise check, it seems to be pretty hard to design it. My suggestion is to only verify properties that you need to process the output. So `kClsOutputName` should just be ignored the else branch.

By the way, it looks that `handwritten-score-recognition-0001` has a batch dimension in inputs but doesn't have it in outputs. That's uncommon. Is it intentional?

Reply:
Alright, I've removed the hardcoding of the `kClsOutputName`.
The model that I've added is horizontal-text-detection-0001 and, indeed, I've just noticed this odd thing. The inputs have the batch structure [1, 3, 704, 704], while outputs have [100, 5] dimension, where 100 - max number of prediction per image. But it is defined architecturally.

Reply:
That probably means the model doesn't work with different batch size. If this is true, I think it's better to point out that model's property in its description.

Reply:
For object detection models it's common to not have a separate batch dimension in the output; instead, the dimension corresponding to the number of objects is multiplied by the batch size. The same could be the case here.

Reply:
@Wovchena Thank you for your time. I've fixed the model's description. Do you like it or you suggest to describe it more clearly?

Reply:
LGTM